### PR TITLE
Add a PAPI data provider that reads scripture using ParatextData

### PR DESCRIPTION
**Pull request by lyonsil**
This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261)

### CODE REVIEW

**Review by tjcouch-sil — COMMENTED** [review]
How exciting!! Looking forward to having this in.

Sorry I forgot to think about this before so we could talk about design, but I lean against making this special external types extension. I think I would prefer if we made an actual extension and used the normal build process instead of adding special code for this types file. Maybe we could go for an extension that gives types for all the C# stuff, or we could go for many extensions that are split up (it is likely that people will not want all the functionality from the C# data provider one day once we provide ways to activate and deactivate extensions, so it may be best to split them up somewhat by feature). I guess it would be worth adding a special thing in `extension.service.ts` that checks if the manifest has a main provided at all and ignores it if not? Or we could add a boolean in the manifest like "onlyTypes" or something that indicates the extension only provides types (so it is clearer that you must provide either `main` or `onlyTypes` and we can error otherwise)? Other people may want to provide TS types that correspond to external code, so it would be good to figure out how we want to do this and make it more official.

It is also likely that there will be some JS code that corresponds to external code in some cases, and it may be that people want to join those together in one extension. Like a FLEx extension that has a JS webview or something. Just a thought for considering how we want to do this.

> Reviewed 3 of 18 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYYg3EJ7HIx37ggZCBY:bndoock)* status: 3 of 18 files reviewed, all discussions resolved

**Review by lyonsil — COMMENTED** [review]
> I think I would prefer if we made an actual extension and used the normal build process instead of adding special code for this types file.

I'm not sure what an "actual extension" would be in this case.  Maybe you just mean the "extension with no code" you talk about later.  If there are cases where an extension is sort of a combination of C# code and JS/TS code, then for sure I think it makes sense to have an actual extension there.  In cases where there is no JS/TS code needed to call the extension, though, I would think we want some way to expose types.

Would it be reasonable to go with this current approach in this PR for now and have a follow up ticket to redesign how to expose types in TS for non-TS data providers?  I feel like there could be a lot of back and forth about this point alone, and Tim might want to chime in.  I'd like to get this merged soon so the extensions team has a clear way to load USFM data to work with.

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYYqRLw10mlQbw3r6uf:b-b7rqr5)* status: 3 of 18 files reviewed, all discussions resolved (waiting on @tjcouch-sil)

**Review by irahopkinson — CHANGES_REQUESTED** [review]
I get this error on the terminal in Windows:


> Reviewed 16 of 18 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NY_KQqP96W7AVm9zaE_:bv7s795)* status: all files reviewed, 5 unresolved discussions (waiting on @lyonsil)

___
*[`cspell.json` line 46 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_Mwfu412sCYN7UOSM:-NY_Mwfu412sCYN7UOSN:b32wwy) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/cspell.json#L46)):*
> ```JSON
>     "unsubscribers",
>     "usfm",
>     "versificaiton"
> ```

sp
___
*[`c-sharp/NetworkObjects/UsfmDataProvider.cs` line 10 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_Mn7L4Y9TG9SReKIY:-NY_Mn7M7Kjf9mPYHso9:bquj3od) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/NetworkObjects/UsfmDataProvider.cs#L10)):*
> ```Smalltalk
> namespace Paranext.DataProvider.NetworkObjects
> {
>     // TODO: Add versificaiton support
> ```

BTW sp



_Code quote:_
```Smalltalk
versificaiton 
```
___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 12 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_NfI37nTQ_Si1iJJF:-NY_NfI37nTQ_Si1iJJG:b-miocku) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L12)):*
> ```Smalltalk
>         private static bool s_initialized = false;
> 
>         public static void Initialize(string dataFolder)
> ```

BTW I'm guessing `dataFolder` is a path - how about renaming to `dataFolderPath` to clarify. This would apply elsewhere also.
___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 20 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_P8FI-B8CYKUr0d1V:-NY_P8FI-B8CYKUr0d1W:bojvhuw) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L20)):*
> ```Smalltalk
>             lock (s_locker)
>             {
>                 if (s_initialized)
> ```

BTW could this check be outside the lock?
___
*[`c-sharp-tests/ParatextDataConnectionTests.cs` line 54 at r2](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_BGym3FFSb5Zi6F_o:-NY_BGym3FFSb5Zi6F_p:b-l9ieu8) ([raw file](https://github.com/paranext/paranext-core/blob/92ee4da8dea81128f65a8d50b6683324a3a4781b/c-sharp-tests/ParatextDataConnectionTests.cs#L54)):*
> ```Smalltalk
> 
>             ScrText scrText = ScrTextCollection.Find("WEB");
>             Assert.Multiple(() =>
> ```

BTW curious why you made this multiple? Seems like failing on the first error in the order given would be appropriate?

**Review by lyonsil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261)* status: 9 of 23 files reviewed, 4 unresolved discussions (waiting on @irahopkinson and @tjcouch-sil)

___
*[`cspell.json` line 46 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_Mwfu412sCYN7UOSM:-NYdfB5j-d89zygaSQ8t:bs2zy69) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/cspell.json#L46)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

sp
</blockquote></details>

Indeed - fixed now
___
*[`c-sharp/NetworkObjects/UsfmDataProvider.cs` line 10 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_Mn7L4Y9TG9SReKIY:-NYdfKZc0feT1ehYsr19:bmgdo4m) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/NetworkObjects/UsfmDataProvider.cs#L10)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

BTW sp
</blockquote></details>

Yep, this was the original problem - fixed
___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 20 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_P8FI-B8CYKUr0d1V:-NYdfQeW0O_Xhw_3cVm6:b-6vppex) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L20)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

BTW could this check be outside the lock?
</blockquote></details>

Moving it outside the lock introduces a small race condition.  I could check outside the lock, too, which is a reasonable pattern people follow: https://en.wikipedia.org/wiki/Double-checked_locking.

It really shouldn't matter for what we're doing here.

**Review by lyonsil — COMMENTED** [review]
This error is fixed in the latest changes.

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYdm4Yp-sgrqNHgBZfz:bqfntc7)* status: 8 of 23 files reviewed, 1 unresolved discussion (waiting on @irahopkinson and @tjcouch-sil)

___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 12 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_NfI37nTQ_Si1iJJF:-NYdhB9mEemst80I6p8T:bwmu070) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L12)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

BTW I'm guessing `dataFolder` is a path - how about renaming to `dataFolderPath` to clarify. This would apply elsewhere also.
</blockquote></details>

It's a relative path, but in this case we're using something in the same directory (i.e., "assets").  Updated the name.
___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 20 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_P8FI-B8CYKUr0d1V:-NYdiXY50ELT5_Sa67D_:bwc5wej) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L20)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

Moving it outside the lock introduces a small race condition.  I could check outside the lock, too, which is a reasonable pattern people follow: https://en.wikipedia.org/wiki/Double-checked_locking.

It really shouldn't matter for what we're doing here.
</blockquote></details>

I went ahead and added the second check.
___
*[`c-sharp-tests/ParatextDataConnectionTests.cs` line 54 at r2](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_BGym3FFSb5Zi6F_o:-NYdg7li-OgoQZoOuvWP:b-o67104) ([raw file](https://github.com/paranext/paranext-core/blob/92ee4da8dea81128f65a8d50b6683324a3a4781b/c-sharp-tests/ParatextDataConnectionTests.cs#L54)):*
<details><summary><i>Previously, irahopkinson (Ira Hopkinson) wrote…</i></summary><blockquote>

BTW curious why you made this multiple? Seems like failing on the first error in the order given would be appropriate?
</blockquote></details>

I made this change based on Visual Studio's prompting.  Here is the URL with more details on the recommendation/warning I saw: https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit2045.md

Basically I believe the idea is that it can be helpful to see everything that failed instead of seeing one issue, fixing it, then not realizing there is another failure right after it that might require a different or additional fix.

**Review by irahopkinson — DISMISSED** [review]
Yep thanks, working for me now on Windows too. FYI I can't get the 2x hello world tabs to open on Linux but I don't think that is due to this PR. I get a few `TypeError: fetchOriginal is not a function` errors. How do you get them showing up?

> Reviewed 13 of 14 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYoY-g84YyRNZOOr6x-:b-n0iicm)* status: all files reviewed, 2 unresolved discussions (waiting on @lyonsil)

___
*[`c-sharp/ParatextUtils/ParatextGlobals.cs` line 20 at r3](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_P8FI-B8CYKUr0d1V:-NYoMBOq8pGAHOAtJDBk:b7qrhsn) ([raw file](https://github.com/paranext/paranext-core/blob/42bc80f9dd69a12c2b4f824df3aece71257350f3/c-sharp/ParatextUtils/ParatextGlobals.cs#L20)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I went ahead and added the second check.
</blockquote></details>

Right, this is trickier than it looks at first with the lock.
___
*[`c-sharp-tests/ParatextDataConnectionTests.cs` line 54 at r2](https://reviewable.io/reviews/paranext/paranext-core/261#-NY_BGym3FFSb5Zi6F_o:-NYoVv507SqDswn4oc2g:b-cyey0t) ([raw file](https://github.com/paranext/paranext-core/blob/92ee4da8dea81128f65a8d50b6683324a3a4781b/c-sharp-tests/ParatextDataConnectionTests.cs#L54)):*
<details><summary><i>Previously, lyonsil (Matt Lyons) wrote…</i></summary><blockquote>

I made this change based on Visual Studio's prompting.  Here is the URL with more details on the recommendation/warning I saw: https://github.com/nunit/nunit.analyzers/blob/master/documentation/NUnit2045.md

Basically I believe the idea is that it can be helpful to see everything that failed instead of seeing one issue, fixing it, then not realizing there is another failure right after it that might require a different or additional fix.
</blockquote></details>

Yes, I had looked that up too since I wasn't familiar with it. I guess it would make more sense to me to use the multiple on the last 2 and move the `Assert.That(scrText, Is.Not.Null);` before it since the last 2 won't have any meaning if `scrText` is null.
___
*[`extensions/lib/external-usfm-data-provider/index.d.ts` line 1 at r5](https://reviewable.io/reviews/paranext/paranext-core/261#-NYoTpHI7b27WwGQg__R:-NYoTpHJ0VNoF6fmWWom:b-dfivic) ([raw file](https://github.com/paranext/paranext-core/blob/5ecce054c3fedf950ab58d4024e2ed82e49b2b1b/extensions/lib/external-usfm-data-provider/index.d.ts#L1)):*
> ```XML
> declare module 'external-usfm-data-provider' {}
> ```

BTW as far as I can see this file isn't really doing anything useful since the declaration is an empty object. However, if it exported items from the adjacent `./usfm-data-provider.d.ts` file then the import in `hello-world.web-view.tsx` could be simplified from `import { type UsfmProviderDataTypes } from '@extensions/external-usfm-data-provider/usfm-data-provider';` to `import { type UsfmProviderDataTypes } from '@extensions/external-usfm-data-provider';`.
___
*[`extensions/vite/vite.config.ts` line 77 at r5](https://reviewable.io/reviews/paranext/paranext-core/261#-NYoNDIzAsx1eYO_e7zP:-NYoNDIzAsx1eYO_e7zQ:bml3u6y) ([raw file](https://github.com/paranext/paranext-core/blob/5ecce054c3fedf950ab58d4024e2ed82e49b2b1b/extensions/vite/vite.config.ts#L77)):*
> ```XML
>       viteStaticCopy({
>         targets: externalExtensions.flatMap<Target>((extension): Target => {
>           return {
> ```

BTW since this only returns a single object without any processing, you can keep this format cleaner by following the example above that uses round brackets so you don't need the extra `{...}` and the `return` inside. So it would look like this:
```typescript
      viteStaticCopy({
        targets: externalExtensions.flatMap<Target>(
          (extension): Target => ({
            src: `${sourceFolder}/${extension}/*.*`,
            dest: `${extension}`,
          }),
        ),
      }),
```

**Review by irahopkinson — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYtxKBvBDEMZQfnQUSr:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @lyonsil)

**Review by irahopkinson — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @lyonsil)

**Review by tjcouch-sil — COMMENTED** [review]
@lyonsil Yes, when I said at the top to make an actual extension, I did mean the same as below where I said basically an extension with no code but that just serves types.

Sure, we can split off the work into a new issue. Good suggestion! :) #267 thanks for considering this with me.

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYu5K3L4PKHa3bQJB5D:b-dnkxwr)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @lyonsil)

File: c-sharp/NetworkObjects/UsfmDataProvider.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            : base("usfm", papiClient)
+        {
+            _collectionName = collectionName;
+            ParatextGlobals.Initialize(dataFolderPath);
+        }
+
+        protected override void StartDataProvider()
+        {
+            _scrText = ScrTextCollection.Find(_collectionName);
+        }
```
Comment (jwickberg) [inline c-sharp/NetworkObjects/UsfmDataProvider.cs:27]: Seems you would want to use FindById to get the project - which would mean that you would need to have id as part of the arguments. Name isn't guaranteed to be unique.

Also, you later give an error about StartDataProvider not being called based on _scrText being null, but the Find and FindById methods will return null if the project isn't found. You probably want a separate error for that.

Comment (lyonsil) [reply c-sharp/NetworkObjects/UsfmDataProvider.cs:27]: For the moment we're loading a project out of fixed data files checked into the repo.  Once we start working with "real" projects with multiple collections, this will definitely need to be reworked.  There isn't a design in play yet for how we will work with projects, so I went with something very simple.

**Review by lyonsil — COMMENTED** [review]
> Dismissed @jwickberg from a discussion.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261)* status: 22 of 23 files reviewed, 1 unresolved discussion (waiting on @FoolRunning, @irahopkinson, @jwickberg, and @tjcouch-sil)

**Review by irahopkinson — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/261#-:-NYyfx-h4Rr_izojbpt7:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @FoolRunning, @jwickberg, and @tjcouch-sil)
